### ISSUE TITLE
Not able to add Keycloak as OIDC

### ISSUE THREAD

**Issue by afw8k** [Stirling-Tools/Stirling-PDF#1561]
Hello, I am not able to add Keycloak as my OIDC Provider. 

this is my docker compose : 

version: '3.3'
services:
  stirling-pdf:
    image: frooodle/s-pdf:latest-fat
    ports:
      - '8080:8080'
    volumes:
      - ./trainingData:/usr/share/tessdata # Required for extra OCR languages
      - ./extraConfigs:/configs
      - ./customFiles:/customFiles/
      - ./logs:/logs/
    environment:
      - DOCKER_ENABLE_SECURITY=false
      - INSTALL_BOOK_AND_ADVANCED_HTML_OPS=false
      - LANGS=de_DE
      - SECURITY_ENABLELOGIN=true
      - SECURITY_CSRFDISABLED=false
      - SECURITY_LOGINATTEMPTCOUNT=5
      - SECURITY_LOGINRESETTIMEMINUTES=120
      - SECURITY_LOGINMETHOD=all
      - SECURITY_INITIALLOGIN_USERNAME=admin
      - SECURITY_INITIALLOGIN_PASSWORD=github
      - SECURITY_OAUTH2_ENABLED=true
      - SECURITY_OAUTH2_CLIENT_KEYCLOAK_ISSUER=https://xx.xx.xx/realms/xxxxx
      - SECURITY_OAUTH2_CLIENT_KEYCLOAK_CLIENTID=stirling
      - SECURITY_OAUTH2_CLIENT_KEYCLOAK_CLIENTSECRET=xxxxxxx
      - SECURITY_OAUTH2_CLIENT_KEYCLOAK_SCOPES=openid, profile, email
      - SECURITY_OAUTH2_CLIENT_KEYCLOAK_USEASUSERNAME=preferred_username
      - UI_APPNAME=xxxxx-PDF
      - UI_HOMEDESCRIPTION=The ultimate PDF solution
      - UI_APPNAMENAVBAR=xxxxx-PDF
      - SYSTEM_DEFAULTLOCALE=de-DE
      - SYSTEM_GOOGLEVISIBILITY=false
      - SYSTEM_ENABLEALPHAFUNCTIONALITY=false
      - SYSTEM_SHOWUPDATE=true
      - SYSTEM_SHOWUPDATEONLYADMIN=false
      - SYSTEM_CUSTOMHTMLFILES=false
      - METRICS_ENABLED=true

this is my keycloak configuration...
![image](https://github.com/user-attachments/assets/efb91a5e-345f-4bde-8a1b-c6e83e07ab11)
![image](https://github.com/user-attachments/assets/f504bd07-3ecd-4821-8fcf-79a2e6a5d92f)

So whats the issue? : 
Stirling-PDF Login Screen appears. I can press on OIDC Option Keycloak. After i get redirected to Keycloak and authentificate myself. after successful Authenification i get this error: 

![image](https://github.com/user-attachments/assets/a3e1409a-a8ff-4979-b47c-1c0277ea3427)

here are the logs of info.log file:

root@xxxx-lrds-dmz-01:/home/xxxx/stirling/logs# cat info.log
2024-07-14 19:46:12,789 INFO s.s.S.SPdfApplication [main] Starting SPdfApplication v0.26.1 using Java 21.0.3 with PID 10 (/app.jar started by stirlingpdfuser in /)
2024-07-14 19:46:12,796 INFO s.s.S.SPdfApplication [main] No active profile set, falling back to 1 default profile: "default"
2024-07-14 19:46:16,708 INFO o.s.d.r.c.RepositoryConfigurationDelegate [main] Bootstrapping Spring Data JPA repositories in DEFAULT mode.
2024-07-14 19:46:16,949 INFO o.s.d.r.c.RepositoryConfigurationDelegate [main] Finished Spring Data repository scanning in 220 ms. Found 3 JPA repository interfaces.
2024-07-14 19:46:19,740 INFO o.s.b.w.e.j.JettyServletWebServerFactory [main] Server initialized with port: 8080
2024-07-14 19:46:19,755 INFO o.e.j.s.Server [main] jetty-12.0.7; built: 2024-02-29T21:19:41.771Z; git: c89aca8fd34083befd79f328a3b8b6ffff04347e; jvm 21.0.3+9-alpine-r0
2024-07-14 19:46:19,833 INFO o.e.j.s.h.C.application [main] Initializing Spring embedded WebApplicationContext
2024-07-14 19:46:19,836 INFO o.s.b.w.s.c.ServletWebServerApplicationContext [main] Root WebApplicationContext: initialization completed in 6827 ms
2024-07-14 19:46:20,716 INFO o.h.j.i.u.LogHelper [main] HHH000204: Processing PersistenceUnitInfo [name: default]
2024-07-14 19:46:20,877 INFO o.h.Version [main] HHH000412: Hibernate ORM core version 6.4.4.Final
2024-07-14 19:46:20,972 INFO o.h.c.i.RegionFactoryInitiator [main] HHH000026: Second-level cache disabled
2024-07-14 19:46:21,714 INFO o.s.o.j.p.SpringPersistenceUnitInfo [main] No LoadTimeWeaver setup: ignoring JPA class transformer
2024-07-14 19:46:21,787 INFO c.z.h.HikariDataSource [main] HikariPool-1 - Starting...
2024-07-14 19:46:22,319 INFO c.z.h.p.HikariPool [main] HikariPool-1 - Added connection conn0: url=jdbc:h2:file:./configs/stirling-pdf-DB user=SA
2024-07-14 19:46:22,325 INFO c.z.h.HikariDataSource [main] HikariPool-1 - Start completed.
2024-07-14 19:46:25,945 INFO o.h.e.t.j.p.i.JtaPlatformInitiator [main] HHH000489: No JTA platform available (set 'hibernate.transaction.jta.platform' to enable JTA platform integration)
2024-07-14 19:46:26,087 INFO o.s.o.j.LocalContainerEntityManagerFactoryBean [main] Initialized JPA EntityManagerFactory for persistence unit 'default'
2024-07-14 19:46:27,712 INFO o.e.j.s.DefaultSessionIdManager [main] Session workerName=node0
2024-07-14 19:46:27,744 INFO o.e.j.s.h.ContextHandler [main] Started osbwej.JettyEmbeddedWebAppContext@2565b097{application,/,b=[file:/tmp/jetty-docbase.8080.14213280410739442792/, URLResource@BA730246(jar:nested:///app.jar/!BOOT-INF/lib/swagger-ui-5.2.0.jar!/META-INF/resources/)],a=AVAILABLE,h=oeje10s.SessionHandler@5fe80760{STARTED}}
2024-07-14 19:46:27,746 INFO o.e.j.e.s.ServletContextHandler [main] Started osbwej.JettyEmbeddedWebAppContext@2565b097{application,/,b=[file:/tmp/jetty-docbase.8080.14213280410739442792/, URLResource@BA730246(jar:nested:///app.jar/!BOOT-INF/lib/swagger-ui-5.2.0.jar!/META-INF/resources/)],a=AVAILABLE,h=oeje10s.SessionHandler@5fe80760{STARTED}}
2024-07-14 19:46:27,776 INFO o.e.j.s.Server [main] Started oejs.Server@401ec794{STARTING}[12.0.7,sto=0] @18228ms
2024-07-14 19:46:27,829 INFO s.s.S.SPdfApplication [main] Running configs ApplicationProperties [security=Security [enableLogin=true, oauth2=OAUTH2 [enabled=true, issuer=null, clientId=null, clientSecret=NULL, autoCreateUser=false, useAsUsername=null, provider=null, scopes=[]], initialLogin=InitialLogin [username=admin, password=MASKED], csrfDisabled=false], system=System [defaultLocale=de-DE, googlevisibility=false, enableAlphaFunctionality=false, showUpdate=true, showUpdateOnlyAdmin=false], ui=UserInterface [appName=xxxx-PDF, homeDescription=The ultimate PDF solution, appNameNavbar=xxxx-PDF], endpoints=Endpoints [toRemove=[], groupsToRemove=[]], metrics=Metrics [enabled=true], automaticallyGenerated=AutomaticallyGenerated [key=MASKED], autoPipeline=null]
2024-07-14 19:46:29,452 INFO s.s.S.c.EndpointConfiguration [main] Disabling pdf-to-book
2024-07-14 19:46:29,453 INFO s.s.S.c.EndpointConfiguration [main] Disabling book-to-pdf
2024-07-14 19:46:31,580 WARN o.s.b.a.o.j.JpaBaseConfiguration$JpaWebConfiguration [main] spring.jpa.open-in-view is enabled by default. Therefore, database queries may be performed during view rendering. Explicitly configure spring.jpa.open-in-view to disable this warning
2024-07-14 19:46:31,836 INFO o.s.b.a.w.s.WelcomePageHandlerMapping [main] Adding welcome page: ServletContext resource [/index.html]
2024-07-14 19:46:32,345 INFO o.s.b.a.e.w.EndpointLinksResolver [main] Exposing 1 endpoint(s) beneath base path '/actuator'
2024-07-14 19:46:32,596 INFO o.s.s.w.DefaultSecurityFilterChain [main] Will secure any request with [org.springframework.security.web.session.DisableEncodeUrlFilter@afee49e, org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter@4c3d056, org.springframework.security.web.context.SecurityContextHolderFilter@159e7734, org.springframework.security.web.header.HeaderWriterFilter@5710598e, org.springframework.web.filter.CorsFilter@346b793b, org.springframework.security.web.authentication.logout.LogoutFilter@27d7ee42, org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestRedirectFilter@77eab063, org.springframework.security.oauth2.client.web.OAuth2LoginAuthenticationFilter@16c30e11, stirling.software.SPDF.config.security.UserAuthenticationFilter@291fba9, stirling.software.SPDF.config.security.IPRateLimitingFilter@3435b6d8, org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter@227051e6, stirling.software.SPDF.config.security.FirstLoginFilter@11bfd751, org.springframework.security.web.session.ConcurrentSessionFilter@5455f457, org.springframework.security.web.savedrequest.RequestCacheAwareFilter@324fb35d, org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter@488897ef, org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter@18fb5226, org.springframework.security.web.authentication.AnonymousAuthenticationFilter@396ea7ab, org.springframework.security.web.session.SessionManagementFilter@5b9ec0, org.springframework.security.web.access.ExceptionTranslationFilter@60339ff4, org.springframework.security.web.access.intercept.AuthorizationFilter@388b4f68]
2024-07-14 19:46:33,562 INFO o.e.j.s.h.C.application [main] Initializing Spring DispatcherServlet 'dispatcherServlet'
2024-07-14 19:46:33,563 INFO o.s.w.s.DispatcherServlet [main] Initializing Servlet 'dispatcherServlet'
2024-07-14 19:46:33,565 INFO o.s.w.s.DispatcherServlet [main] Completed initialization in 2 ms
2024-07-14 19:46:33,588 INFO o.e.j.s.AbstractConnector [main] Started ServerConnector@536d655f{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
2024-07-14 19:46:33,590 INFO o.s.b.w.e.j.JettyWebServer [main] Jetty started on port 8080 (http/1.1) with context path '/'
2024-07-14 19:46:33,621 INFO s.s.S.SPdfApplication [main] Started SPdfApplication in 22.114 seconds (process running for 24.073)
2024-07-14 19:46:33,635 WARN s.s.S.u.FileMonitor [scheduling-1] not monitoring any directory, even the root directory itself: ./pipeline/watchedFolders
2024-07-14 19:46:33,639 INFO s.s.S.u.FileMonitor [scheduling-1] Registered directory: ./pipeline/watchedFolders
2024-07-14 19:46:34,637 INFO s.s.S.SPdfApplication [main] Stirling-PDF Started.
2024-07-14 19:46:34,638 INFO s.s.S.SPdfApplication [main] Navigate to http://localhost:8080

the file invalid-auths.log is empty. 

i use nginx reverse proxy manager. 

Thank you for your kind help.

**Comment by Ludy87** [comment]
Hi @afw8k I couldn't reproduce the problem. Your configuration is fine as far as I can see.

**Comment by AhmedK2151** [comment]
Hi, I haven't used Keycloak but I see that your Redirect URI field doesn't match what is written in the [Stirling-PDF docs for SSO](https://stirlingtools.com/docs/Advanced%20Configuration/Single%20Sign-On%20Configuration). Maybe that could be the issue.

This is what my Redirect URI field looks like in Authentik for reference: `https://<domain>/login/oauth2/code/oidc`.
This configuration works without issue.

One issue that I did have was that I included `/.well-known/openid-configuration` in the `SECURITY_OAUTH2_CLIENT_ISSUER` field.  This resulted in the app crashing.

I hope this helps.
